src/parser/jobs/mch/modules/Reassemble.tsx

```diff
@@ -1,86 +1,108 @@
-import {Trans, Plural} from '@lingui/react'
-import {ActionLink} from 'components/ui/DbLink'
-import {ActionRoot} from 'data/ACTIONS/root'
-import STATUSES from 'data/STATUSES'
-import {CastEvent, BuffEvent} from 'fflogs'
-import Module, {dependency} from 'parser/core/Module'
-import Combatants from 'parser/core/modules/Combatants'
+import {Plural, Trans} from '@lingui/react'
+import {DataLink} from 'components/ui/DbLink'
+import {ActionKey} from 'data/ACTIONS'
+import {Event, Events} from 'event'
+import {Analyser} from 'parser/core/Analyser'
+import {filter, oneOf} from 'parser/core/filter'
+import {dependency} from 'parser/core/Injectable'
 import {Data} from 'parser/core/modules/Data'
-import Suggestions, {TieredSuggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
 import React from 'react'
 
 // These are the only GCDs that should be reassembled under normal circumstances
-const REASSEMBLE_GCDS: Array<keyof ActionRoot> = [
+const REASSEMBLE_GCDS: ActionKey[] = [
 	'AIR_ANCHOR',
 	'AUTO_CROSSBOW',
 	'DRILL',
 	'SPREAD_SHOT',
+	// 6.0:
+	//'CHAIN_SAW',
+	//'SCATTER_GUN',
 ]
 
-const OTHER_GCDS: Array<keyof ActionRoot> = [
+const OTHER_GCDS: ActionKey[] = [
 	'BIOBLASTER',
 	'HEAT_BLAST',
 	'HEATED_SPLIT_SHOT',
 	'HEATED_SLUG_SHOT',
 	'HEATED_CLEAN_SHOT',
 ]
 
-export default class Reassemble extends Module {
+export class Reassemble extends Analyser {
 	static override handle = 'reassemble'
 
-	@dependency private combatants!: Combatants
 	@dependency private data!: Data
 	@dependency private suggestions!: Suggestions
 
 	private badReassembles = 0
 	private droppedReassembles = 0
-	private lastUse = 0
+	private lastGcd = 0
+	private reassembleActive = false
 
 	private REASSEMBLE_GCDS: number[] = []
 	private OTHER_GCDS: number[] = []
 
-	protected override init() {
+	override initialise() {
 		this.REASSEMBLE_GCDS = REASSEMBLE_GCDS.map(key => this.data.actions[key].id)
 		this.OTHER_GCDS = OTHER_GCDS.map(key => this.data.actions[key].id)
 
-		this.addEventHook('cast', {by: 'player', abilityId: [...this.REASSEMBLE_GCDS, ...this.OTHER_GCDS]}, this.onCast)
-		this.addEventHook('removedebuff', {by: 'player', abilityId: STATUSES.REASSEMBLED.id}, this.onRemoveReassemble)
+		const reassembleFilter = filter<Event>()
+			.source(this.parser.actor.id)
+			.status(this.data.statuses.REASSEMBLED.id)
+
+		const gcdFilter = filter<Event>()
+			.source(this.parser.actor.id)
+			.type('action')
+			.action(oneOf([...this.REASSEMBLE_GCDS, ...this.OTHER_GCDS]))
+
+		this.addEventHook(reassembleFilter.type('statusApply'), this.onReassemble)
+		this.addEventHook(gcdFilter, this.onCast)
```

Comment:
Is it really worth hooking every GCD cast for quite a large list? Consider adding the hook in `onReassemble` and removing it in `onRemove`. Functionality inside the callback shouldn't need to change, but you'll be hooking way less junk events since `lastGcd` only matters inside an active Reassemble status. Since it's consumed by GCDs, timing derpery between pressing Reassemble and getting the status shouldn't be an issue.

Reply:
I'm not going to push hard for this, but yes. Adding/removing hooks is cheap, but every active hook has a slight overhead. If the data used by a hook is not used outside defineable windows, it's preferable to only hook it during those windows.